File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/Delphi/ProviderModeDemo/ProviderModeDemoClasses.pas
unit ProviderModeDemoClasses;

{$I cxVer.inc}

interface

uses
  Classes, cxTL, cxTLData, cxCustomData;

type
  { TcxProviderRecordHandle }

  TcxProviderRecordHandle = class
  private
    FChildLevelCount: Integer;
    FChildList: TList;
    FDate: TDateTime;
    FDataLoaded: Boolean;
    FDeletion: Boolean;
    FIntValue: Integer;
    FParent: TcxProviderRecordHandle;
    FText: string;
    function GetCount: Integer;
    function GetItem(Index: Integer): TcxProviderRecordHandle;
    function GetIndex: Integer;
    function GetLevel: Integer;
    function GetTotalCount: Integer;
  protected
    property Deletion: Boolean read FDeletion;
    property Parent: TcxProviderRecordHandle read FParent;
  public
    constructor Create(AParent: TcxProviderRecordHandle);
    destructor Destroy; override;
    function Add: TcxProviderRecordHandle;
    function AddChild: TcxProviderRecordHandle;
    procedure DeleteChildren;
    procedure NodeMoveTo(AttachRecordHandle: TcxProviderRecordHandle;
      AttachMode: TcxTreeListNodeAttachMode; IsCopy: Boolean);
    property ChildLevelCount: Integer read FChildLevelCount;
    property Count: Integer read GetCount;
    property Index: Integer read GetIndex;
    property IntValue: Integer read FIntValue write FIntValue;
    property Items[Index: Integer]: TcxProviderRecordHandle read GetItem; default;
    property Date: TDateTime read FDate write FDate;
    property Level: Integer read GetLevel;
    property Text: string read FText write FText;
    property TotalCount: Integer read GetTotalCount;
  end;

  { TcxCustomDemoDataSource}

  TcxCustomDemoDataSource = class(TcxTreeListCustomDataSource)
  private
    FID: Integer; 
    FRootHandle: TcxProviderRecordHandle;
    function InsertRecordHandle(AParentHandle: TcxProviderRecordHandle;
      AIsChild: Boolean): TcxProviderRecordHandle; virtual;
    procedure GenerateChildRecords(AParentHandle: TcxProviderRecordHandle);
  protected
    procedure NodeMoveTo(ARecordHandle, AttachRecordHandle: TcxDataRecordHandle;
      AttachMode: TcxTreeListNodeAttachMode; IsCopy: Boolean); override;
    function AppendRecord: TcxDataRecordHandle; override;
    procedure DeleteRecord(ARecordHandle: TcxDataRecordHandle); override;
    function GetValue(ARecordHandle: TcxDataRecordHandle; AItemHandle: TcxDataItemHandle): Variant; override;
    function InsertRecord(ARecordHandle: TcxDataRecordHandle): TcxDataRecordHandle; override;
    procedure SetValue(ARecordHandle: TcxDataRecordHandle; AItemHandle: TcxDataItemHandle; const AValue: Variant); override;
    property RootHandle: TcxProviderRecordHandle read FRootHandle;
  public
    constructor Create; virtual;
    destructor Destroy; override;
  end;

  TcxCustomDemoDataSourceClass  = class of TcxCustomDemoDataSource;

  { TcxSmartLoadDemoDataSource }

  TcxSmartLoadDemoDataSource = class(TcxCustomDemoDataSource)
  protected
    function AppendRecord: TcxDataRecordHandle; override;
    function GetChildCount(AParentHandle: TcxDataRecordHandle): Integer; override;
    function GetChildRecordHandle(AParentHandle: TcxDataRecordHandle;
      AChildIndex: Integer): TcxDataRecordHandle; override;
    function GetRootRecordHandle: TcxDataRecordHandle; override;
  public
    constructor Create; override;
  end;

  { TcxLoadAllRecordsDemoDataSource }

  TcxLoadAllRecordsDemoDataSource = class(TcxCustomDemoDataSource)
  private
    FRecordsList: TList;
    procedure CreateAllRecords;
    function InsertRecordHandle(AParentHandle: TcxProviderRecordHandle;
      AIsChild: Boolean): TcxProviderRecordHandle; override;
  protected
    procedure DeleteRecord(ARecordHandle: TcxDataRecordHandle); override;
    function GetParentRecordHandle(ARecordHandle: TcxDataRecordHandle): TcxDataRecordHandle; override;
    function GetRecordHandle(ARecordIndex: Integer): TcxDataRecordHandle; override;
    function GetRecordCount: Integer; override;
  public
    constructor Create; override;
    destructor Destroy; override;
  end;

  procedure RecreateDemoDataSource(ATreeList: TcxVirtualTreeList);

implementation

uses
  SysUtils, Dialogs;

const
  cxProviderDemoLevelCount = 5;
  cxProviderDemoRecordsPerLevel = 10;


procedure RecreateDemoDataSource(ATreeList: TcxVirtualTreeList);
var
  ADataSource: TObject;
const
  AClasses: array[Boolean] of TClass =
    (TcxLoadAllRecordsDemoDataSource, TcxSmartLoadDemoDataSource);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  ADataSource := ATreeList.CustomDataSource;
  ATreeList.CustomDataSource := nil;
  FreeAndNil(ADataSource);
  ATreeList.CustomDataSource :=
    TcxCustomDemoDataSourceClass(AClasses[ATreeList.OptionsData.SmartLoad]).Create;
//}

end;


{ TcxProviderRecordHandle }

constructor TcxProviderRecordHandle.Create(
  AParent: TcxProviderRecordHandle);
begin
  FParent := AParent;
  FChildList := TList.Create;
  if AParent <> nil then
    Parent.FChildList.Add(Self);
  if AParent <> nil then
    FChildLevelCount := AParent.ChildLevelCount - 1
  else
    FChildLevelCount := cxProviderDemoLevelCount;
end;

destructor TcxProviderRecordHandle.Destroy;
begin
  try
    DeleteChildren;
  finally
    FChildList.Free;
    if (Parent <> nil) and not Parent.Deletion then
      Parent.FChildList.Remove(Self);
    inherited Destroy;
  end;
end;

function TcxProviderRecordHandle.Add: TcxProviderRecordHandle;
begin
  Result := TcxProviderRecordHandle.Create(Parent);
end;

function TcxProviderRecordHandle.AddChild: TcxProviderRecordHandle;
begin
  Result := TcxProviderRecordHandle.Create(Self);
end;

procedure TcxProviderRecordHandle.DeleteChildren;
var
  I: Integer;
begin
  FDeletion := True;
  try
    for I := 0 to Count - 1 do
      TObject(FChildList.List[I]).Free;
  finally
    FChildList.Clear;
    FDeletion := False;
  end;
end;

procedure TcxProviderRecordHandle.NodeMoveTo(
  AttachRecordHandle: TcxProviderRecordHandle;
  AttachMode: TcxTreeListNodeAttachMode; IsCopy: Boolean);
  procedure ChangeParent(AParent: TcxProviderRecordHandle);
  begin
    if FParent <> AParent then
    begin
      FParent.FChildList.Remove(Self);
      AParent.FChildList.Add(Self);
      FParent := AParent;
    end;
  end;
begin
  case AttachMode of
    tlamAdd, tlamAddFirst, tlamInsert:
      ChangeParent(AttachRecordHandle.Parent);
    tlamAddChild, tlamAddChildFirst:
      ChangeParent(AttachRecordHandle);
  end;
end;

function TcxProviderRecordHandle.GetCount: Integer;
begin
  Result := FChildList.Count;
end;

function TcxProviderRecordHandle.GetItem(Index: Integer): TcxProviderRecordHandle;
begin
  Result := TcxProviderRecordHandle(FChildList[Index]);
end;

function TcxProviderRecordHandle.GetIndex: Integer;
begin
  if Parent <> nil then
    Result := Parent.FChildList.IndexOf(Self)
  else
    Result := -1; 
end;

function TcxProviderRecordHandle.GetLevel: Integer;
var
  AParent: TcxProviderRecordHandle;
begin
  Result := -1;
  AParent := Parent;
  while AParent <> nil do
  begin
    AParent := AParent.Parent;
    Inc(Result);
  end;
end;

function TcxProviderRecordHandle.GetTotalCount: Integer;
var
  I: Integer;
begin
  Result := Count;
  for I := 0 to Count - 1 do
    Inc(Result, Items[I].TotalCount);
end;

{ TcxCustomDemoDataSource }

constructor TcxCustomDemoDataSource.Create;
begin
  FRootHandle := TcxProviderRecordHandle.Create(nil);
end;

destructor TcxCustomDemoDataSource.Destroy;
begin
  FRootHandle.Free;
  inherited Destroy;
end;

function TcxCustomDemoDataSource.AppendRecord: TcxDataRecordHandle;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Result := InsertRecordHandle(RootHandle, True);
  TcxProviderRecordHandle(Result).FDataLoaded := True;
  DataChanged;

//}
end;

procedure TcxCustomDemoDataSource.DeleteRecord(ARecordHandle: TcxDataRecordHandle);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  TcxProviderRecordHandle(ARecordHandle).Free;
  DataChanged;

//}
end;

function TcxCustomDemoDataSource.GetValue(ARecordHandle: TcxDataRecordHandle;
  AItemHandle: TcxDataItemHandle): Variant;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  with TcxProviderRecordHandle(ARecordHandle) do
    case Integer(AItemHandle) of
      0: Result := IntValue;
      1: Result := Text;
      2: Result := Date;
      3:
        if Parent = nil then
          Result := -1
        else
          Result := Parent.IntValue;
    end;

//}
end;

function TcxCustomDemoDataSource.InsertRecord(
  ARecordHandle: TcxDataRecordHandle): TcxDataRecordHandle;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Result := InsertRecordHandle(ARecordHandle, False);
  TcxProviderRecordHandle(Result).FDataLoaded := True;
  DataChanged;

//}
end;

procedure TcxCustomDemoDataSource.SetValue(
  ARecordHandle: TcxDataRecordHandle; AItemHandle: TcxDataItemHandle;
  const AValue: Variant);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  with TcxProviderRecordHandle(ARecordHandle) do
    case Integer(AItemHandle) of
      0: IntValue := AValue;
      1: Text := AValue;
      2: Date := AValue;
    end;

//
end;

procedure TcxCustomDemoDataSource.NodeMoveTo(ARecordHandle,
  AttachRecordHandle: TcxDataRecordHandle;
  AttachMode: TcxTreeListNodeAttachMode; IsCopy: Boolean);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  if IsCopy then
    with InsertRecordHandle(TcxProviderRecordHandle(AttachRecordHandle),
      AttachMode in [tlamAddChild, tlamAddChildFirst]) do
    begin
      IntValue := TcxProviderRecordHandle(ARecordHandle).IntValue;
      Date := TcxProviderRecordHandle(ARecordHandle).Date;
      Text := TcxProviderRecordHandle(ARecordHandle).Text;
    end
  else
    TcxProviderRecordHandle(ARecordHandle).NodeMoveTo(
      TcxProviderRecordHandle(AttachRecordHandle), AttachMode, IsCopy);
  DataChanged;

//}
end;

function TcxCustomDemoDataSource.InsertRecordHandle(
  AParentHandle: TcxProviderRecordHandle; AIsChild: Boolean): TcxProviderRecordHandle;
begin
  Inc(FID);
  if AIsChild then
    Result := AParentHandle.AddChild
  else
    Result := AParentHandle.Add
end;

procedure TcxCustomDemoDataSource.GenerateChildRecords(
  AParentHandle: TcxProviderRecordHandle);
var
  I: Integer;
begin
  for I := 0 to cxProviderDemoRecordsPerLevel - 1 do
    with InsertRecordHandle(AParentHandle, True) do
    begin
      IntValue := FID;
      Text := 'Text' + IntToStr(FIntValue);
      Date := Now + FIntValue * 0.001;
    end;
  AParentHandle.FDataLoaded := True;
end;

{ TcxSmartLoadDemoDataSource }

constructor TcxSmartLoadDemoDataSource.Create;
begin
  inherited Create;
  GenerateChildRecords(FRootHandle);
end;

function TcxSmartLoadDemoDataSource.AppendRecord: TcxDataRecordHandle;
var
  AIndex: Integer;
  AHandle: TcxDataRecordHandle;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  AIndex := DataController.FocusedRecordIndex;
  if AIndex = -1 then
    AHandle := RootHandle
  else
    AHandle := TcxProviderRecordHandle(GetRecordHandleByIndex(AIndex)).Parent;
  Result := InsertRecordHandle(AHandle, True);
  TcxProviderRecordHandle(Result).FDataLoaded := True;
  DataChanged;

//}
end;

function TcxSmartLoadDemoDataSource.GetChildCount(
  AParentHandle: TcxDataRecordHandle): Integer;

  function GetCountFromItem(AItem: TcxProviderRecordHandle): Integer;
  begin
    Result := AItem.Count;
    if not AItem.FDataLoaded then
    begin
      if (AItem.ChildLevelCount > 0) then
        Result := Result + cxProviderDemoRecordsPerLevel
      else
        AItem.FDataLoaded := True;
    end;
  end;

begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Result := GetCountFromItem(TcxProviderRecordHandle(AParentHandle));

//}
end;

function TcxSmartLoadDemoDataSource.GetChildRecordHandle(
  AParentHandle: TcxDataRecordHandle; AChildIndex: Integer): TcxDataRecordHandle;

  function GetChildItemHandle(AItem: TcxProviderRecordHandle): TcxDataRecordHandle;
  begin
    if not AItem.FDataLoaded then
      GenerateChildRecords(AItem);
    Result := AItem.Items[AChildIndex]
  end;

begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Result := GetChildItemHandle(TcxProviderRecordHandle(AParentHandle));

//}
end;

function TcxSmartLoadDemoDataSource.GetRootRecordHandle: TcxDataRecordHandle;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Result := FRootHandle;

//}
end;

{ TcxLoadAllRecordsDemoDataSource }

constructor TcxLoadAllRecordsDemoDataSource.Create;
begin
  inherited Create;
  FRecordsList := TList.Create;
  CreateAllRecords; 
end;

destructor TcxLoadAllRecordsDemoDataSource.Destroy;
begin
  FRecordsList.Free;
  inherited Destroy;
end;

procedure TcxLoadAllRecordsDemoDataSource.DeleteRecord(ARecordHandle: TcxDataRecordHandle);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  FRecordsList.Remove(ARecordHandle);
  inherited DeleteRecord(ARecordHandle);

//}
end;

function TcxLoadAllRecordsDemoDataSource.GetParentRecordHandle(
  ARecordHandle: TcxDataRecordHandle): TcxDataRecordHandle;
begin
  Result := TcxProviderRecordHandle(ARecordHandle).Parent; 
end;

function TcxLoadAllRecordsDemoDataSource.GetRecordHandle(
  ARecordIndex: Integer): TcxDataRecordHandle;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Result := FRecordsList[ARecordIndex];

//}
end;

function TcxLoadAllRecordsDemoDataSource.GetRecordCount: Integer;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  Result := FRecordsList.Count;

//}
end;

function TcxLoadAllRecordsDemoDataSource.InsertRecordHandle(
  AParentHandle: TcxProviderRecordHandle; AIsChild: Boolean): TcxProviderRecordHandle;
begin
  Result := inherited InsertRecordHandle(AParentHandle, AIsChild);
  FRecordsList.Add(Result);
end;

procedure TcxLoadAllRecordsDemoDataSource.CreateAllRecords;

   procedure DoCreateRecords(AParent: TcxProviderRecordHandle; ALevel: Integer);
   var
     I: Integer;
   begin
     if ALevel > cxProviderDemoLevelCount then Exit;
     GenerateChildRecords(AParent);
     for I := 0 to AParent.Count - 1 do
       DoCreateRecords(AParent.Items[I], ALevel + 1);
   end;
begin
  DoCreateRecords(RootHandle, 1);
end;

end.

File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/CBuilder/ProviderModeDemo/ProviderModeDemoClasses.cpp
//---------------------------------------------------------------------------


#pragma hdrstop

#include "ProviderModeDemoClasses.h"

//---------------------------------------------------------------------------

#pragma package(smart_init)

//---------------------------------------------------------------------------

void RecreateDemoDataSource(TcxVirtualTreeList *ATreeList)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  TObject *ADataSource = ATreeList->CustomDataSource;
  ATreeList->CustomDataSource = NULL;
  if (ADataSource != NULL)
    delete ADataSource;
  if (ATreeList->OptionsData->SmartLoad)
    ATreeList->CustomDataSource = new TcxSmartLoadDemoDataSource();
  else
    ATreeList->CustomDataSource = new TcxLoadAllRecordsDemoDataSource();

//*/
}

//---------------------------------------------------------------------------

__fastcall TcxProviderRecordHandle::TcxProviderRecordHandle(
  TcxProviderRecordHandle *AParent)
{
  FParent = AParent;
  FDataLoaded = false;
  FChildList = new TList();
  if (AParent != NULL)
	Parent->AddExistChild(this);
  if (AParent != NULL)
	FChildLevelCount = AParent->ChildLevelCount - 1;
  else
	FChildLevelCount = cxProviderDemoLevelCount;
}
//---------------------------------------------------------------------------

__fastcall TcxProviderRecordHandle::~TcxProviderRecordHandle()
{
  __try{
	DeleteChildren();
  }
  __finally{
	delete FChildList;
    FChildList = NULL;
	if ((Parent != NULL) && !Parent->Deletion)
	  Parent->RemoveChild(this);
  }
}
//---------------------------------------------------------------------------

TcxProviderRecordHandle* __fastcall TcxProviderRecordHandle::Add()
{
  return new TcxProviderRecordHandle(Parent);
}
//---------------------------------------------------------------------------

TcxProviderRecordHandle* __fastcall TcxProviderRecordHandle::AddChild()
{
  return new TcxProviderRecordHandle(this);
}
//---------------------------------------------------------------------------

void __fastcall TcxProviderRecordHandle::AddExistChild(
  TcxProviderRecordHandle* ARecordHandle)
{
  FChildList->Add(ARecordHandle);
}
//---------------------------------------------------------------------------

void __fastcall TcxProviderRecordHandle::DeleteChildren()
{
  FDeletion = true;
  __try{
	for (int I = Count - 1; I = 0 ; I--)
	  delete (TcxProviderRecordHandle*)FChildList->Items[I]; 
  }
  __finally{
	FChildList->Clear();
	FDeletion = false;
  }
}
//---------------------------------------------------------------------------

int __fastcall TcxProviderRecordHandle::GetChildIndex(
  TcxProviderRecordHandle* ARecordHandle)
{
  return FChildList->IndexOf(ARecordHandle);
}
//---------------------------------------------------------------------------

void __fastcall TcxProviderRecordHandle::NodeMoveTo(
  TcxProviderRecordHandle *AttachRecordHandle,
  TcxTreeListNodeAttachMode AttachMode, bool IsCopy)
{
  switch (AttachMode) {
    case tlamAdd:
    case tlamAddFirst:
    case tlamInsert:
      ChangeParent(AttachRecordHandle->Parent);
      break;
    case tlamAddChild:
    case tlamAddChildFirst:
      ChangeParent(AttachRecordHandle);
      break;
  }
}

void __fastcall TcxProviderRecordHandle::ChangeParent(TcxProviderRecordHandle *AParent)
{
  if (FParent != AParent){
    FParent->RemoveChild(this);
    AParent->AddExistChild(this);
    FParent = AParent;
  }
}
//---------------------------------------------------------------------------

void __fastcall TcxProviderRecordHandle::RemoveChild(
  TcxProviderRecordHandle* ARecordHandle)
{
  FChildList->Remove(ARecordHandle);
}
//---------------------------------------------------------------------------

int __fastcall TcxProviderRecordHandle::GetCount()
{
  return FChildList->Count;
}
//---------------------------------------------------------------------------

TcxProviderRecordHandle* __fastcall TcxProviderRecordHandle::GetItem(int Index)
{
  return (TcxProviderRecordHandle*)FChildList->Items[Index];
}
//---------------------------------------------------------------------------

int __fastcall TcxProviderRecordHandle::GetIndex()
{
  if (Parent != NULL)
    return Parent->GetChildIndex(this);
  else
    return -1;
}
//---------------------------------------------------------------------------

int __fastcall TcxProviderRecordHandle::GetLevel()
{
  int Result = -1;
  TcxProviderRecordHandle *AParent = Parent;
  while (AParent != NULL){
    AParent = AParent->Parent;
    Result ++;
  }
  return Result;
}
//---------------------------------------------------------------------------

int __fastcall TcxProviderRecordHandle::GetTotalCount()
{
  int Result = Count;
  for (int I = 0; I < Count; I++)
    Result += Items[I]->TotalCount;
  return Result;
}
//---------------------------------------------------------------------------


/*TcxCustomDemoDataSource*/

__fastcall TcxCustomDemoDataSource::TcxCustomDemoDataSource(void)
{
  FRootHandle = new TcxProviderRecordHandle(NULL);
}
//---------------------------------------------------------------------------

__fastcall TcxCustomDemoDataSource::~TcxCustomDemoDataSource(void)
{
  delete FRootHandle;
}
//---------------------------------------------------------------------------

void * __fastcall TcxCustomDemoDataSource::AppendRecord(void)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  TcxDataRecordHandle Result = InsertRecordHandle(RootHandle, true);
  ((TcxProviderRecordHandle*)Result)->DataLoaded = true;
  DataChanged();
  return Result;

//*/
}
//---------------------------------------------------------------------------

void __fastcall TcxCustomDemoDataSource::DeleteRecord(void * ARecordHandle)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  delete (TcxProviderRecordHandle*)ARecordHandle;
  DataChanged();

//*/
}
//---------------------------------------------------------------------------

Variant __fastcall TcxCustomDemoDataSource::GetValue(
  void * ARecordHandle, void * AItemHandle)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  TcxProviderRecordHandle *AHandle = (TcxProviderRecordHandle*)ARecordHandle;
  Variant Result;
  switch ((int)AItemHandle){
    case 0:
      Result = AHandle->IntValue;
      break;
    case 1:
      Result = AHandle->Text;
      break;
    case 2:
      Result = AHandle->Date;
      break;
    case 3:
      if (AHandle->Parent == NULL)
        Result = -1;
      else
        Result = AHandle->Parent->IntValue;
  }
  return Result;

//*/
}
//---------------------------------------------------------------------------

void * __fastcall TcxCustomDemoDataSource::InsertRecord(void * ARecordHandle)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  TcxDataRecordHandle Result =
    InsertRecordHandle((TcxProviderRecordHandle*)ARecordHandle, false);
  ((TcxProviderRecordHandle*)Result)->DataLoaded = true;
  DataChanged();
  return Result;

//*/
}
//---------------------------------------------------------------------------

TcxProviderRecordHandle* __fastcall TcxCustomDemoDataSource::InsertRecordHandle(
  TcxProviderRecordHandle *AParentHandle, bool AIsChild)
{
  FID++;
  TcxProviderRecordHandle* Result;
  if (AIsChild)
    Result = AParentHandle->AddChild();
  else
    Result = AParentHandle->Add();
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall TcxCustomDemoDataSource::NodeMoveTo(void * ARecordHandle,
  void * AttachRecordHandle, Cxtl::TcxTreeListNodeAttachMode AttachMode, bool IsCopy)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  if (IsCopy){
    TcxProviderRecordHandle* AProviderRecordHandle =
      InsertRecordHandle((TcxProviderRecordHandle*)AttachRecordHandle,
        (AttachMode == tlamAddChild || AttachMode == tlamAddChildFirst));
    AProviderRecordHandle->IntValue =
      ((TcxProviderRecordHandle*)ARecordHandle)->IntValue;
    AProviderRecordHandle->Date = ((TcxProviderRecordHandle*)ARecordHandle)->Date;
    AProviderRecordHandle->Text = ((TcxProviderRecordHandle*)ARecordHandle)->Text;
  }
  else
    ((TcxProviderRecordHandle*)ARecordHandle)->NodeMoveTo(
      (TcxProviderRecordHandle*)AttachRecordHandle, AttachMode, IsCopy);
  DataChanged();

//*/
}
//---------------------------------------------------------------------------

void __fastcall TcxCustomDemoDataSource::SetValue(void * ARecordHandle,
  void * AItemHandle, const Variant &AValue)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  TcxProviderRecordHandle *ProviderRecordHandle =
    (TcxProviderRecordHandle*)ARecordHandle;
  switch ((int)AItemHandle){
    case 0:
      ProviderRecordHandle->IntValue = AValue;
      break;
    case 1:
      ProviderRecordHandle->Text = AValue;
      break;
    case 2:
      ProviderRecordHandle->Date = VarToDateTime(AValue);
  }

//*/
}
//---------------------------------------------------------------------------

void __fastcall TcxCustomDemoDataSource::GenerateChildRecords(
  TcxProviderRecordHandle *AParentHandle)
{
  for (int I = 0; I < cxProviderDemoRecordsPerLevel; I++){
    TcxProviderRecordHandle *ProviderRecordHandle =
      InsertRecordHandle(AParentHandle, true);
    ProviderRecordHandle->IntValue = FID;
    ProviderRecordHandle->Text = "Text" + IntToStr(ProviderRecordHandle->IntValue);
    ProviderRecordHandle->Date = Now() + ProviderRecordHandle->IntValue * 0.001;
  }
  AParentHandle->DataLoaded = true;
}
//---------------------------------------------------------------------------

/*TcxSmartLoadDemoDataSource */

__fastcall TcxSmartLoadDemoDataSource::TcxSmartLoadDemoDataSource(void):TcxCustomDemoDataSource()
{
  GenerateChildRecords(FRootHandle);
}
//---------------------------------------------------------------------------

void * __fastcall TcxSmartLoadDemoDataSource::AppendRecord(void)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  int AIndex = DataController->FocusedRecordIndex;
  TcxDataRecordHandle Result;
  if (AIndex == -1)
    Result = RootHandle;
  else
    Result = ((TcxProviderRecordHandle*)GetRecordHandleByIndex(AIndex))->Parent;
  Result = InsertRecordHandle((TcxProviderRecordHandle*)Result, true);
  ((TcxProviderRecordHandle*)Result)->DataLoaded = true;
  DataChanged();
  return Result;

//*/
}
//---------------------------------------------------------------------------

int __fastcall TcxSmartLoadDemoDataSource::GetChildCount(
  void * AParentHandle)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  return GetCountFromItem((TcxProviderRecordHandle*)AParentHandle);

//*/
}

int __fastcall TcxSmartLoadDemoDataSource::GetCountFromItem(
  TcxProviderRecordHandle *AItem)
{
  int Result = AItem->Count;
  if (!AItem->DataLoaded){
    if (AItem->ChildLevelCount > 0)
      Result += cxProviderDemoRecordsPerLevel;
    else
      AItem->DataLoaded = true;
  }
  return Result;
}
//---------------------------------------------------------------------------

void * __fastcall TcxSmartLoadDemoDataSource::GetChildRecordHandle(
  void * AParentHandle, int AChildIndex)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  return GetChildItemHandle((TcxProviderRecordHandle*)AParentHandle, AChildIndex);

//*/
}

TcxDataRecordHandle __fastcall TcxSmartLoadDemoDataSource::GetChildItemHandle(
  TcxProviderRecordHandle *AItem, int AChildIndex)
{
  if (!AItem->DataLoaded)
    GenerateChildRecords(AItem);
  return AItem->Items[AChildIndex];
}
//---------------------------------------------------------------------------

void * __fastcall TcxSmartLoadDemoDataSource::GetRootRecordHandle(void)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  return FRootHandle;

//*/
}
//---------------------------------------------------------------------------


/*TcxLoadAllRecordsDemoDataSource */

__fastcall TcxLoadAllRecordsDemoDataSource::TcxLoadAllRecordsDemoDataSource(void):
  TcxCustomDemoDataSource()
{
  FRecordsList = new TList;
  CreateAllRecords();
}
//---------------------------------------------------------------------------

__fastcall TcxLoadAllRecordsDemoDataSource::~TcxLoadAllRecordsDemoDataSource(void)
{
  delete FRecordsList;
}
//---------------------------------------------------------------------------

void __fastcall TcxLoadAllRecordsDemoDataSource::DeleteRecord(
  void * ARecordHandle)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  FRecordsList->Remove(ARecordHandle);
  TcxCustomDemoDataSource::DeleteRecord(ARecordHandle);

//*/
}
//---------------------------------------------------------------------------

int __fastcall TcxLoadAllRecordsDemoDataSource::GetRecordCount(void)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  return FRecordsList->Count;

//*/
}
//---------------------------------------------------------------------------

void * __fastcall TcxLoadAllRecordsDemoDataSource::GetParentRecordHandle(void * ARecordHandle)
{
  return ((TcxProviderRecordHandle*)ARecordHandle)->Parent;
}
//---------------------------------------------------------------------------

void * __fastcall TcxLoadAllRecordsDemoDataSource::GetRecordHandle(
  int ARecordIndex)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  return FRecordsList->Items[ARecordIndex];

//*/
}
//---------------------------------------------------------------------------

TcxProviderRecordHandle* __fastcall TcxLoadAllRecordsDemoDataSource::InsertRecordHandle(
  TcxProviderRecordHandle *AParentHandle, bool AIsChild)
{
  TcxProviderRecordHandle *Result =
    TcxCustomDemoDataSource::InsertRecordHandle(AParentHandle, AIsChild);
  FRecordsList->Add(Result);
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall TcxLoadAllRecordsDemoDataSource::CreateAllRecords()
{
  DoCreateRecords(RootHandle, 1);
}

void __fastcall TcxLoadAllRecordsDemoDataSource::DoCreateRecords(
  TcxProviderRecordHandle *AParent, int ALevel)
{
  if (ALevel > cxProviderDemoLevelCount) return;
  GenerateChildRecords(AParent);
  for (int I = 0; I < AParent->Count; I++)
    DoCreateRecords(AParent->Items[I], ALevel + 1);
}
//---------------------------------------------------------------------------

